<a href="https://colab.research.google.com/github/EstherYanLi/Customer-Churn-Prediction/blob/main/Customer_churn_prediction_using_updated_IBM_data_Telco_Github.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1.Introduction

In the new version of IBM Cognos data set 'The Telco customer churn' contains information about a fictional telco company that provided home phone and Internet services to 7043 customers in California in Q3. It indicates which customers have left, stayed, or signed up for their service. This data set was enhanced to provide a wider narrative. Multiple important demographics are included for each customer, as well as a Churn Score and Customer Lifetime Value (CLTV) index.

Explain of important variables?


Telco customer churn (11.1.3+):
https://community.ibm.com/community/user/businessanalytics/blogs/steven-macko/2019/07/11/telco-customer-churn-1113


# 2.Importing librabries




In [ ]:
# Update numpy to the latest version to resolve potential binary incompatibility issues
!pip install --upgrade numpy


In [ ]:
# Uninstall and reinstall scikit-learn to resolve numpy binary incompatibility
!pip uninstall -y scikit-learn
!pip install scikit-learn

# Load libraries
import pandas as pd  # For data manipulation and analysis
import numpy as np   # For numerical computations
import matplotlib.pyplot as plt  # For data visualization
import seaborn as sns  # For advanced data visualization

from sklearn import preprocessing, model_selection, metrics  # For machine learning
from sklearn.preprocessing import OneHotEncoder, StandardScaler  # For preprocessing data
from sklearn.model_selection import train_test_split  # For splitting data into training and testing sets
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report  # For evaluating model performance

from sklearn.linear_model import LogisticRegression  # For logistic regression
from sklearn.svm import SVC # For SVC
from sklearn.ensemble import RandomForestClassifier  # For random forest classification

  Using cached scikit_learn-1.8.0-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (11 kB)
Using cached scikit_learn-1.8.0-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (8.9 MB)


# 3.Data loading

In [ ]:
# Mount drive
from google.colab import drive
drive.mount('/content/drive')

# Read the file from Google Drive
df = pd.read_csv("/content/drive/My Drive/Work/Colab Notebooks/Dataset/Copy of Telco_customer_churn.csv")


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 4.Exploratory data analysis EDA

In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7032 non-null   object 
 1   Senior Citizen     7032 non-null   object 
 2   Partner            7032 non-null   object 
 3   Dependents         7032 non-null   object 
 4   Tenure Months      7032 non-null   int64  
 5   Phone Service      7032 non-null   object 
 6   Multiple Lines     7032 non-null   object 
 7   Internet Service   7032 non-null   object 
 8   Online Security    7032 non-null   object 
 9   Online Backup      7032 non-null   object 
 10  Device Protection  7032 non-null   object 
 11  Tech Support       7032 non-null   object 
 12  Streaming TV       7032 non-null   object 
 13  Streaming Movies   7032 non-null   object 
 14  Contract           7032 non-null   object 
 15  Paperless Billing  7032 non-null   object 
 16  Payment Method     7032 non-n

In [ ]:
df.head(5)

,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,...,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Value,Churn Score
0,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,...,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,1,86
1,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,...,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,1,67
2,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,...,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.50,1,86
3,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,...,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,1,84
4,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,...,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.30,1,89


*It is noticed that 'Total charges' is object which should be floating data type.*

In [ ]:
#statistical summary
df.describe()

In [ ]:
df.isnull().sum()

Churn Reason has 5174 missing data.

In [ ]:
# Understand Why 'Churn Reason' has 5174 missing data, Check the Relationship Between Churn and Churn Reason
# Cross-tabulate 'Churn Label' and 'Churn Reason' missingness
print(pd.crosstab(df['Churn Label'], df['Churn Reason'].isnull(), margins=True))

"Churn Reason" is structurally missing for retained customers (Churn="No"), creating data leakage since missingness itself reveals customer status. We recommend remove the column to eliminates leakage risk.

# 5.Data Visualization

## 5.1 Numerical data with churn value

In [ ]:
# Plot numeric variables with respect to Churn label
numeric_vars = ['Tenure Months', 'Monthly Charges', 'Churn Score', 'CLTV']

fig, axes = plt.subplots(nrows=2, ncols=3, figsize=(18, 12))
axes = axes.flatten()

for i, var in enumerate(numeric_vars):
    sns.boxplot(x='Churn Label', y=var, data=df, ax=axes[i])
    axes[i].set_title(f'{var} Distribution by Churn Label')

# Remove empty subplots (if any)
for j in range(len(numeric_vars), len(axes)):
    fig.delaxes(axes[j])

plt.tight_layout()
plt.show()


## 5.2 Catergorical data with churn value

In [ ]:
# Plot categorical variables with respect to Churn label
categorical_vars = ['Gender', 'Senior Citizen', 'Partner', 'Dependents',
       'Phone Service', 'Multiple Lines', 'Internet Service',
       'Online Security', 'Online Backup', 'Device Protection', 'Tech Support',
       'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing',
       'Payment Method']

fig, axes = plt.subplots(nrows=6, ncols=3, figsize=(18, 36))
axes = axes.flatten()

for i, var in enumerate(categorical_vars):
    sns.countplot(x=var, hue='Churn Label', data=df, palette='viridis', ax=axes[i])
    axes[i].set_title(f'{var} by Churn Label')
    axes[i].set_xlabel(var)
    axes[i].set_ylabel('Count')
    axes[i].tick_params(axis='x', rotation=45)
    axes[i].legend(title='Churn Label', loc='upper right')

# Remove empty subplots (if any)
for j in range(len(categorical_vars), len(axes)):
    fig.delaxes(axes[j])

plt.tight_layout()
plt.show()

In [ ]:
# Calculate the correlation matrix
numeric_df = df[numeric_vars]
corr = numeric_df.corr()

# Plot heatmap
plt.figure(figsize=(8, 6))
sns.heatmap(data=corr,annot=True,fmt=".2f",cmap="coolwarm",linewidths=0.5,vmin=-1,vmax=1)
plt.title("Correlation Matrix")
plt.show()

# 6.Data Preprocessing

## 6.1 Dropping unnecessary columns

In [ ]:
# Delete columns if not necessary
columns_to_drop = ['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Churn Reason', 'CLTV', 'Churn Label']
df = df.drop(columns=[col for col in columns_to_drop if col in df.columns], axis=1)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7043 non-null   object 
 1   Senior Citizen     7043 non-null   object 
 2   Partner            7043 non-null   object 
 3   Dependents         7043 non-null   object 
 4   Tenure Months      7043 non-null   int64  
 5   Phone Service      7043 non-null   object 
 6   Multiple Lines     7043 non-null   object 
 7   Internet Service   7043 non-null   object 
 8   Online Security    7043 non-null   object 
 9   Online Backup      7043 non-null   object 
 10  Device Protection  7043 non-null   object 
 11  Tech Support       7043 non-null   object 
 12  Streaming TV       7043 non-null   object 
 13  Streaming Movies   7043 non-null   object 
 14  Contract           7043 non-null   object 
 15  Paperless Billing  7043 non-null   object 
 16  Payment Method     7043 

## 6.2 Converting catergorical to numerical variables

In [ ]:
# Since 'Total Charges' is object, we need to convert 'Total Charges' column to numeric, coercing errors to NaN
df['Total Charges'] = pd.to_numeric(df['Total Charges'], errors='coerce')

# Check for missing values in the DataFrame
print(df.isnull().sum())

Gender                0
Senior Citizen        0
Partner               0
Dependents            0
Tenure Months         0
Phone Service         0
Multiple Lines        0
Internet Service      0
Online Security       0
Online Backup         0
Device Protection     0
Tech Support          0
Streaming TV          0
Streaming Movies      0
Contract              0
Paperless Billing     0
Payment Method        0
Monthly Charges       0
Total Charges        11
Churn Value           0
Churn Score           0
dtype: int64


## 6.3 Handling missing values

In [ ]:
# Delete rows with missing values in 'Total Charges' column
df=df.dropna(subset=['Total Charges'])
print(df.isnull().sum())

Gender               0
Senior Citizen       0
Partner              0
Dependents           0
Tenure Months        0
Phone Service        0
Multiple Lines       0
Internet Service     0
Online Security      0
Online Backup        0
Device Protection    0
Tech Support         0
Streaming TV         0
Streaming Movies     0
Contract             0
Paperless Billing    0
Payment Method       0
Monthly Charges      0
Total Charges        0
Churn Value          0
Churn Score          0
dtype: int64


## 6.4 Splitting the data

In [ ]:
# Split the data
X = df.drop('Churn Value', axis=1)
y = df['Churn Value'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 6.5 Encoding Categorical Variables

In [ ]:
# Get categorical features
cat_features = X_train.select_dtypes(include=['object']).columns.tolist()

# OneHotEncoding for Nominal Categorical Features
encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')  # handle_unknown='ignore' to avoid errors during inference
X_train_ohe = encoder.fit_transform(X_train[cat_features])
X_test_ohe = encoder.transform(X_test[cat_features])

# Create DataFrames for encoded data
X_train_ohe_df = pd.DataFrame(X_train_ohe, columns=encoder.get_feature_names_out(cat_features), index=X_train.index)
X_test_ohe_df = pd.DataFrame(X_test_ohe, columns=encoder.get_feature_names_out(cat_features), index=X_test.index)

# Concatenate encoded features with numerical features
X_train = pd.concat([X_train.drop(cat_features, axis=1), X_train_ohe_df], axis=1)
X_test = pd.concat([X_test.drop(cat_features, axis=1), X_test_ohe_df], axis=1)

## 6.6 Standardization

In [ ]:
# Standardization
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 7.Building models and evaluation

## 7.1 Logistic Regression model

In [ ]:
# Train a Logistic Regression model
model_lg = LogisticRegression(random_state=42)
model_lg.fit(X_train, y_train)

# Make predictions
y_pred = model_lg.predict(X_test)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

## 7.2 SVM model

In [ ]:
# Train a SVM model
model_svm = SVC()
model_svm.fit(X_train, y_train)

# Make predictions
y_pred = model_svm.predict(X_test)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

## 7.3 Random Forest Classifier

In [ ]:
# Train a Random Forest Classifier
model_rf = RandomForestClassifier(n_estimators=100 , oob_score = True, n_jobs = -1,
                                  random_state =50, max_features = 20,
                                  max_leaf_nodes = 30)
model_rf.fit(X_train, y_train)

# Make predictions
y_pred = model_rf.predict(X_test)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

## 7.4 Feature Importance Analysis for Random Forest Classifier

Let's analyze the feature importances to understand which features contribute most to the Random Forest model's predictions.

In [ ]:
# Get feature names (numerical features that remained after initial drops, before one-hot encoding)
numerical_features_after_drop = X.select_dtypes(include=['int64', 'float64']).columns.tolist()

# Get feature names after one-hot encoding for categorical features
categorical_feature_names_ohe = encoder.get_feature_names_out(cat_features)

# Combine all feature names in the correct order as used in training
feature_names = numerical_features_after_drop + list(categorical_feature_names_ohe)

# Get feature importances from the trained Random Forest model
importances = model_rf.feature_importances_

# Create a Series for easier handling and sorting
feature_importances = pd.Series(importances, index=feature_names)

# Sort features by importance
sorted_feature_importances = feature_importances.sort_values(ascending=False)

# Plotting the top N feature importances
plt.figure(figsize=(12, 8))
sns.barplot(x=sorted_feature_importances.head(20).values, y=sorted_feature_importances.head(20).index, palette='viridis')
plt.title('Top 20 Feature Importances (Random Forest)')
plt.xlabel('Importance')
plt.ylabel('Feature')
plt.tight_layout()
plt.show()

Interpretation of Feature Importances

The bar plot above displays the top 20 features ranked by their importance to the Random Forest model. A higher importance score indicates that the feature had a greater impact on the model's predictions.

From this visualization, you can observe:

*   **`Churn Score`** is overwhelmingly the most important feature. This is expected given its direct relationship to churn, acting almost as a proxy for the target variable itself. While highly predictive, in a real-world scenario, you might not have access to a 'Churn Score' directly before a customer churns. If this were the case, you would likely remove this feature to build a predictive model based on other customer attributes.
*   **`Contract_Month-to-month`**, **`Tenure Months`**, **`Online Security_No`**, and **`Total Charges`** are among the next most important features, indicating that contract type, length of service, and lack of online security are significant drivers of churn. `Monthly Charges` also appears to be important.
*   Many one-hot encoded categorical features (e.g., `Gender`, `Dependents`, various service options) have much lower importance scores, suggesting they contribute less to the model's predictive power compared to the top features.

Simplifying the Model by Removing Less Important Features

To simplify the model, we can select only the most important features and re-train the model. This can lead to a more interpretable model, potentially faster training times, and sometimes even improved generalization by reducing noise from irrelevant features. Let's try retraining the Random Forest model using only the top 10 features identified by the `feature_importances`.

In [ ]:
# Select the top N most important features
top_n_features = 15
selected_features = sorted_feature_importances.head(top_n_features).index.tolist()

print(f"Selected Top {top_n_features} Features: {selected_features}")

# Reconstruct the original X (before scaling and one-hot encoding) with selected features
# This requires careful handling of numerical and categorical features

# Identify numerical features among the selected ones
selected_numerical_features = [f for f in selected_features if f in X.select_dtypes(include=['int64', 'float64']).columns.tolist()]

# Identify categorical features (pre-one-hot encoding names) among the selected ones
# This is a bit more complex as one-hot encoded features expand from original categorical features.
# We need to find the original categorical features whose one-hot encoded versions are in selected_features.
selected_categorical_base_features = []
for original_cat_feature in cat_features:
    # Check if any one-hot encoded column derived from this original feature is in selected_features
    if any(f.startswith(original_cat_feature + '_') for f in selected_features):
        selected_categorical_base_features.append(original_cat_feature)

# Create a new DataFrame with only the selected base features (numerical and original categorical)
X_selected_base = df[selected_numerical_features + selected_categorical_base_features]
y_selected = df['Churn Value'].values

# Split the data with the selected base features
X_train_selected, X_test_selected, y_train_selected, y_test_selected = train_test_split(X_selected_base, y_selected, test_size=0.2, random_state=42)

# One-hot encode the selected categorical features
encoder_selected = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
X_train_selected_ohe = encoder_selected.fit_transform(X_train_selected[selected_categorical_base_features])
X_test_selected_ohe = encoder_selected.transform(X_test_selected[selected_categorical_base_features])

X_train_selected_ohe_df = pd.DataFrame(X_train_selected_ohe, columns=encoder_selected.get_feature_names_out(selected_categorical_base_features), index=X_train_selected.index)
X_test_selected_ohe_df = pd.DataFrame(X_test_selected_ohe, columns=encoder_selected.get_feature_names_out(selected_categorical_base_features), index=X_test_selected.index)

# Concatenate encoded features with numerical features
X_train_processed = pd.concat([X_train_selected.drop(selected_categorical_base_features, axis=1), X_train_selected_ohe_df], axis=1)
X_test_processed = pd.concat([X_test_selected.drop(selected_categorical_base_features, axis=1), X_test_selected_ohe_df], axis=1)

# Ensure that the columns in X_train_processed and X_test_processed are exactly the same and in the same order
# This is crucial because selected_features contains the *post-encoding* names.
# We need to filter X_train_processed and X_test_processed to only include columns from `selected_features`.

# Align columns - make sure both train and test have exactly the 'selected_features' in order.
# This involves handling cases where a selected_feature might not be present in the processed X (e.g., due to handle_unknown='ignore' on a rare category not seen in train)

final_selected_features_for_training = [f for f in selected_features if f in X_train_processed.columns]

X_train_processed = X_train_processed[final_selected_features_for_training]
X_test_processed = X_test_processed[final_selected_features_for_training]

# Standardization
scaler_selected = StandardScaler()
X_train_scaled = scaler_selected.fit_transform(X_train_processed)
X_test_scaled = scaler_selected.transform(X_test_processed)

# Re-train Random Forest with selected features
model_rf_simplified = RandomForestClassifier(n_estimators=100 , oob_score = True, n_jobs = -1,
                                             random_state =50, max_features = 20,
                                             max_leaf_nodes = 30)
model_rf_simplified.fit(X_train_scaled, y_train_selected)

# Make predictions with the simplified model
y_pred_simplified = model_rf_simplified.predict(X_test_scaled)

# Evaluate the simplified model
print("\n--- Simplified Random Forest Model Evaluation ---")
print("Accuracy:", accuracy_score(y_test_selected, y_pred_simplified))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_selected, y_pred_simplified))
print("\nClassification Report:\n", classification_report(y_test_selected, y_pred_simplified))

### Analysis of Simplified Model Performance

Compare the evaluation metrics (accuracy, precision, recall, F1-score) of the simplified Random Forest model with the original one. You'll likely observe that:







*   **If the performance is similar or slightly improved**: This indicates that the removed features were indeed less important or even introduced noise, and you have successfully simplified the model without significant performance loss.
*   **If the performance drops significantly**: This suggests that some of the removed features, despite having lower individual importance scores, contributed meaningfully to the model's predictive power (perhaps through complex interactions not fully captured by simple feature importance metrics). In this case, you might need to reconsider the number of features to keep or explore different feature selection techniques.

This process helps you strike a balance between model complexity and performance, leading to a more efficient and understandable model.

### Understanding the Increase in False Negatives (FN)

YouYou've noted that while the overall accuracy remained similar (93%), the False Negatives (FN) increased from 46 to 55 in the simplified model. This is a crucial point, especially in the context of churn prediction.

*   **False Negatives (FN)**: These are customers who *actually churned* but your model *predicted would not churn*. In business terms, these are missed opportunities for intervention. For every FN, you failed to identify a customer at risk, meaning you couldn't implement retention strategies (e.g., offering discounts, personalized support) that might have saved them.

*   **Implication for Churn Prediction**: An increase in FN means your simplified model is slightly less sensitive to detecting actual churners. While overall accuracy is a general measure, in churn prediction, minimizing FN is often a higher priority than minimizing False Positives (FP). This is because the cost of a missed churner (FN) is typically higher (lost revenue, negative word-of-mouth) than the cost of incorrectly identifying a non-churner (FP) (e.g., offering a discount to a loyal customer who wouldn't have churned anyway).

### Error Analysis Using Recall, Precision, and F1-Score

These metrics provide a more nuanced view of your model's performance beyond simple accuracy, especially when dealing with imbalanced classes or when certain types of errors (like FN) are more costly.

*   **Recall (Sensitivity)**: `Recall = True Positives / (True Positives + False Negatives)`
    *   Recall measures the proportion of *actual churners* that your model correctly identified. An increase in FN directly leads to a decrease in recall for the 'churn' class (which is usually your positive class, `1`).
    *   **For churn prediction, high recall is often paramount.** You want to catch as many potential churners as possible, even if it means a slightly higher number of False Positives.

*   **Precision**: `Precision = True Positives / (True Positives + False Positives)`
    *   Precision measures the proportion of *predicted churners* that actually churned. An increase in False Positives (FP) would lead to a decrease in precision.
    *   While high recall is important, precision ensures that your retention efforts are not wasted on too many customers who were never going to churn.

*   **F1-Score**: `F1-Score = 2 * (Precision * Recall) / (Precision + Recall)`
    *   The F1-Score is the harmonic mean of precision and recall. It provides a single metric that balances both. It is particularly useful when you need to strike a balance between precision and recall, or when there's an uneven class distribution.
    *   If your F1-score for the churn class is lower for the simplified model, it suggests a worse balance between precision and recall, often due to the increased FN.

In your case, the original model had 46 FN and the simplified model has 55 FN. This means the recall for the churn class of the simplified model will be slightly lower. Let's look at the classification reports to see the exact values for recall, precision, and F1-score for both models.

In [ ]:
print("--- Original Random Forest Model Classification Report ---")
print(classification_report(y_test, y_pred))

print("\n--- Simplified Random Forest Model Classification Report ---")
print(classification_report(y_test_selected, y_pred_simplified))

## 7.5 Deeper Error Analysis: Examining Misclassified Instances

To understand *why* the model made more False Negatives, you can examine the characteristics of those specific customers. This can reveal patterns or features that the simplified model might be overlooking.

Let's extract and display the False Negative and False Positive instances from your simplified model's predictions.

In [ ]:
# Reconstruct X_test_processed with original indices to easily join back to df
X_test_processed_with_index = pd.DataFrame(X_test_processed, columns=final_selected_features_for_training, index=X_test_selected.index)

# Add actual and predicted labels to the test set for easier analysis
results_df = df.loc[X_test_selected.index].copy()
results_df['Actual_Churn'] = y_test_selected
results_df['Predicted_Churn'] = y_pred_simplified

# Identify False Negatives (Actual Churn = 1, Predicted Churn = 0)
false_negatives = results_df[(results_df['Actual_Churn'] == 1) & (results_df['Predicted_Churn'] == 0)]
print("\n--- False Negative Cases (Actual Churn = Yes, Predicted Churn = No) ---")
print(f"Number of False Negatives: {len(false_negatives)}")
display(false_negatives.head())

# Identify False Positives (Actual Churn = 0, Predicted Churn = 1)
false_positives = results_df[(results_df['Actual_Churn'] == 0) & (results_df['Predicted_Churn'] == 1)]
print("\n--- False Positive Cases (Actual Churn = No, Predicted Churn = Yes) ---")
print(f"Number of False Positives: {len(false_positives)}")
display(false_positives.head())

### How to Use This Information for Improvement:

1.  **Analyze False Negatives**: Look closely at the `false_negatives` DataFrame. Are there any common characteristics among these customers (e.g., similar contract types, service bundles, tenure ranges) that might differentiate them from the True Positives? This could indicate that the simplified model struggles with a particular segment of churners.
2.  **Adjust `top_n_features`**: Experiment with increasing `top_n_features` (e.g., to 20 or 25) in the previous cell to see if including slightly more features helps reduce FN without significantly increasing complexity or FP.
3.  **Targeted Feature Engineering**: If you identify specific patterns in FN, you might consider creating new features or transforming existing ones to better capture those nuances.
4.  **Hyperparameter Tuning (for Recall)**: When tuning the Random Forest, you can specifically optimize for recall (or F1-score) for the positive class, rather than just overall accuracy. This involves using `scoring='recall'` or `scoring='f1'` (for the positive class) in `GridSearchCV` or `RandomizedSearchCV`.
5.  **Class Weights**: Since churn is often an imbalanced problem (fewer churners than non-churners), you can use `class_weight='balanced'` in the `RandomForestClassifier` to give more importance to the minority class during training. This often helps improve recall for the churn class.
6.  **Probability Threshold Adjustment**: For classification models that output probabilities, you can adjust the decision threshold. Lowering the threshold for classifying as 'churn' will generally increase recall (catching more actual churners) but might decrease precision (also identifying more non-churners as churners).

## 7.6 Experiment: Retraining Without 'Churn Score'

As discussed, 'Churn Score' can be considered a form of data leakage as it's highly predictive of churn and likely derived from information that wouldn't be available before a customer actually churns. Let's remove this feature to build a more robust and realistic predictive model.

In [ ]:
# Create a copy of the preprocessed DataFrame to avoid modifying the previous experiments
df_realistic = df.copy()

# Drop 'Churn Score' column for a realistic scenario
df_realistic = df_realistic.drop('Churn Score', axis=1)

print("DataFrame after dropping 'Churn Score':")
df_realistic.info()

DataFrame after dropping 'Churn Score':
<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Gender             7032 non-null   object 
 1   Senior Citizen     7032 non-null   object 
 2   Partner            7032 non-null   object 
 3   Dependents         7032 non-null   object 
 4   Tenure Months      7032 non-null   int64  
 5   Phone Service      7032 non-null   object 
 6   Multiple Lines     7032 non-null   object 
 7   Internet Service   7032 non-null   object 
 8   Online Security    7032 non-null   object 
 9   Online Backup      7032 non-null   object 
 10  Device Protection  7032 non-null   object 
 11  Tech Support       7032 non-null   object 
 12  Streaming TV       7032 non-null   object 
 13  Streaming Movies   7032 non-null   object 
 14  Contract           7032 non-null   object 
 15  Paperless Billing  7032 non-null   ob

In [ ]:
# Split data into features (X) and target (y) for the realistic scenario
X_realistic = df_realistic.drop('Churn Value', axis=1)
y_realistic = df_realistic['Churn Value'].values

# Split into training and testing sets
X_train_realistic, X_test_realistic, y_train_realistic, y_test_realistic = train_test_split(X_realistic, y_realistic, test_size=0.2, random_state=42)

In [ ]:
# Identify categorical features for the realistic scenario
cat_features_realistic = X_train_realistic.select_dtypes(include=['object']).columns.tolist()

# OneHotEncoding for Nominal Categorical Features
encoder_realistic = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
X_train_realistic_ohe = encoder_realistic.fit_transform(X_train_realistic[cat_features_realistic])
X_test_realistic_ohe = encoder_realistic.transform(X_test_realistic[cat_features_realistic])

# Create DataFrames for encoded data
X_train_realistic_ohe_df = pd.DataFrame(X_train_realistic_ohe, columns=encoder_realistic.get_feature_names_out(cat_features_realistic), index=X_train_realistic.index)
X_test_realistic_ohe_df = pd.DataFrame(X_test_realistic_ohe, columns=encoder_realistic.get_feature_names_out(cat_features_realistic), index=X_test_realistic.index)

# Concatenate encoded features with numerical features
X_train_realistic_processed = pd.concat([X_train_realistic.drop(cat_features_realistic, axis=1), X_train_realistic_ohe_df], axis=1)
X_test_realistic_processed = pd.concat([X_test_realistic.drop(cat_features_realistic, axis=1), X_test_realistic_ohe_df], axis=1)

# Standardization
scaler_realistic = StandardScaler()
X_train_realistic_scaled = scaler_realistic.fit_transform(X_train_realistic_processed)
X_test_realistic_scaled = scaler_realistic.transform(X_test_realistic_processed)

print("Data preprocessing for realistic scenario complete.")

Data preprocessing for realistic scenario complete.


In [ ]:
# Re-train Random Forest without 'Churn Score'
model_rf_no_churn_score = RandomForestClassifier(n_estimators=100 , oob_score = True, n_jobs = -1,
                                             random_state =50, max_features = 20,
                                             max_leaf_nodes = 30)
model_rf_no_churn_score.fit(X_train_realistic_scaled, y_train_realistic)

# Make predictions
y_pred_no_churn_score = model_rf_no_churn_score.predict(X_test_realistic_scaled)

# Evaluate the model
print("\n--- Random Forest Model Evaluation (without 'Churn Score') ---")
print("Accuracy:", accuracy_score(y_test_realistic, y_pred_no_churn_score))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_no_churn_score))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_no_churn_score))

### Impact of Removing 'Churn Score'

Now, compare the performance metrics (Accuracy, Precision, Recall, F1-score) of this model against the previous ones. You will likely observe a noticeable drop in performance, particularly in accuracy and recall for the churn class.

This drop is expected, as 'Churn Score' was a very strong predictor. The new model, while potentially less accurate, is a more realistic representation of what you could achieve in a real-world application where such a direct churn indicator is not available. This serves as a baseline for further improvements using only actionable features.

In [ ]:
# Get feature names for the model without Churn Score
numerical_features_no_churn_score = X_realistic.select_dtypes(include=['int64', 'float64']).columns.tolist()

categorical_feature_names_ohe_realistic = encoder_realistic.get_feature_names_out(cat_features_realistic)

feature_names_no_churn_score = numerical_features_no_churn_score + list(categorical_feature_names_ohe_realistic)

# Get feature importances from the trained Random Forest model without 'Churn Score'
importances_no_churn_score = model_rf_no_churn_score.feature_importances_

# Create a Series for easier handling and sorting
feature_importances_no_churn_score = pd.Series(importances_no_churn_score, index=feature_names_no_churn_score)

# Sort features by importance
sorted_feature_importances_no_churn_score = feature_importances_no_churn_score.sort_values(ascending=False)

# Plotting the top N feature importances
plt.figure(figsize=(12, 8))
sns.barplot(x=sorted_feature_importances_no_churn_score.head(20).values, y=sorted_feature_importances_no_churn_score.head(20).index, palette='viridis')
plt.title('Top 20 Feature Importances (Random Forest - No Churn Score)')
plt.xlabel('Importance')
plt.ylabel('Feature')
plt.tight_layout()
plt.show()


### New Feature Importance Analysis (without 'Churn Score')

With 'Churn Score' removed, the feature importance plot now highlights the truly actionable features that drive churn. You can see how `Contract_Month-to-month`, `Tenure Months`, `Online Security_No`, `Internet Service_Fiber optic`, and `Monthly Charges` become the most prominent predictors. This gives a clearer picture of which business drivers to focus on for churn prevention strategies.

## 7.7 Hyperparameter Tuning and Addressing Class Imbalance

### Explaining the Accuracy Drop and Strategies for Improvement

The drop in accuracy was anticipated. The 'Churn Score' was a very strong, almost direct, indicator of churn. Its removal means our model is now relying solely on features that would realistically be available for a customer *before* they churn, which is what we want for a predictive model.

An accuracy of 0.81 is a solid starting point, but there are several ways to improve this **without reintroducing data leakage** or immediately jumping to a new algorithm:

1.  **Hyperparameter Tuning**: The current Random Forest model is using default or basic hyperparameters. Optimizing these parameters can significantly boost performance. We can use techniques like `GridSearchCV` or `RandomizedSearchCV` to find the best combination of parameters.
2.  **Addressing Class Imbalance**: Churn datasets are often imbalanced (fewer churners than non-churners). This can lead models to be biased towards the majority class, making them less effective at identifying the minority class (churners). Strategies include:
    *   **Class Weights**: Adjusting `class_weight` in the `RandomForestClassifier` to 'balanced' or 'balanced_subsample' gives more importance to the minority class.
    *   **Resampling Techniques**: Oversampling the minority class (e.g., SMOTE) or undersampling the majority class.
3.  **Advanced Feature Engineering**: With the 'Churn Score' removed, new combinations or transformations of existing features might become more powerful predictors.
4.  **Experimenting with Other Algorithms**: If after tuning and imbalance handling, the Random Forest doesn't meet the desired performance, then exploring other robust classification algorithms like XGBoost, LightGBM, or even simple neural networks would be the next logical step.

Let's start by trying to improve the current Random Forest model using hyperparameter tuning with `GridSearchCV` and incorporating `class_weight='balanced'` to address potential imbalance.

In [ ]:
from sklearn.model_selection import GridSearchCV

print("Starting Hyperparameter Tuning for Random Forest (without 'Churn Score')")

# Define the parameter grid to search
param_grid = {
    'n_estimators': [100, 200, 300], # Number of trees in the forest
    'max_features': ['sqrt', 'log2'], # Number of features to consider when looking for the best split
    'max_depth': [10, 20, 30, None], # Maximum depth of the tree
    'min_samples_split': [2, 5, 10], # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 2, 4],   # Minimum number of samples required to be at a leaf node
    'class_weight': ['balanced', None] # Handle class imbalance
}

# Initialize the Random Forest Classifier
rf_classifier = RandomForestClassifier(random_state=42, n_jobs=-1, oob_score=True)

# Initialize GridSearchCV
# We'll optimize for 'recall' for the churn class (positive class, which is 1)
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid,
                           cv=3, n_jobs=-1, verbose=2, scoring='recall')

# Fit GridSearchCV
# Ensure X_train_realistic_scaled has the correct column names for consistent feature handling
# If it's a numpy array, we need to convert it back to DataFrame to ensure consistent feature names during scaling if needed.
# For now, assuming X_train_realistic_scaled is already in a format compatible with the model.
grid_search.fit(X_train_realistic_scaled, y_train_realistic)

print("\nBest parameters found:", grid_search.best_params_)
print("Best recall score found:", grid_search.best_score_)

# Get the best model
best_rf_model = grid_search.best_estimator_

# Make predictions with the best model
y_pred_tuned = best_rf_model.predict(X_test_realistic_scaled)

# Evaluate the tuned model
print("\n--- Tuned Random Forest Model Evaluation (without 'Churn Score') ---")
print("Accuracy:", accuracy_score(y_test_realistic, y_pred_tuned))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_tuned))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_tuned))

# 8.Exploring XGBoost Classifier

Now that we've tuned the Random Forest, let's explore other robust classification algorithms, starting with XGBoost. XGBoost (eXtreme Gradient Boosting) is an optimized distributed gradient boosting library designed to be highly efficient, flexible, and portable. It's often a top performer in machine learning competitions due to its ability to handle various data types, missing values, and complex relationships.

We will train an XGBoost classifier, taking into account the class imbalance, and evaluate its performance against the tuned Random Forest.

In [ ]:
import xgboost as xgb

print("Starting XGBoost Classifier Training (without 'Churn Score')")

# Initialize the XGBoost Classifier
# Using 'scale_pos_weight' to handle class imbalance, which is equivalent to 'class_weight'
# Calculate the ratio of negative class to positive class
neg_count = sum(y_train_realistic == 0)
pos_count = sum(y_train_realistic == 1)
scale_pos_weight_value = neg_count / pos_count

xgb_classifier = xgb.XGBClassifier(
    objective='binary:logistic',
    eval_metric='logloss',
    use_label_encoder=False, # Suppress the warning
    random_state=42,
    n_jobs=-1,
    scale_pos_weight=scale_pos_weight_value # Handle class imbalance
)

# Train the model
xgb_classifier.fit(X_train_realistic_scaled, y_train_realistic)

# Make predictions
y_pred_xgb = xgb_classifier.predict(X_test_realistic_scaled)

# Evaluate the model
print("\n--- XGBoost Model Evaluation (without 'Churn Score') ---")
print("Accuracy:", accuracy_score(y_test_realistic, y_pred_xgb))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_xgb))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_xgb))

### Evaluating the XGBoost Model

Compare the performance of this XGBoost model with the tuned Random Forest. Pay close attention to:

*   **Recall for Churn (class 1)**: Is it better than the tuned Random Forest?
*   **Precision for Churn (class 1)**: How does the trade-off between precision and recall compare?
*   **F1-Score for Churn (class 1)**: Does XGBoost provide a better balance?

This will help us determine if XGBoost offers a significant improvement for our realistic churn prediction scenario.

# 9.Exploring LightGBM Classifier

Following XGBoost, let's explore another powerful gradient boosting framework: LightGBM (Light Gradient Boosting Machine). LightGBM is designed to be highly efficient and scalable, especially for large datasets. It often outperforms XGBoost in terms of training speed and can achieve comparable predictive accuracy.

We will train a LightGBM classifier, similar to XGBoost, ensuring we handle class imbalance by setting `is_unbalance=True` or `scale_pos_weight` in its parameters. We will then evaluate its performance and compare it to the Random Forest and XGBoost models.

In [ ]:
import lightgbm as lgb

print("Starting LightGBM Classifier Training (without 'Churn Score')")

# Initialize the LightGBM Classifier
# Using 'scale_pos_weight' to handle class imbalance
# The calculation for neg_count and pos_count is already available from XGBoost step.

lgbm_classifier = lgb.LGBMClassifier(
    objective='binary',
    metric='binary_logloss',
    is_unbalance=True, # Automatically handles imbalanced dataset
    # Alternatively, use scale_pos_weight=scale_pos_weight_value, calculated previously
    random_state=42,
    n_jobs=-1
)

# Train the model
lgbm_classifier.fit(X_train_realistic_scaled, y_train_realistic)

# Make predictions
y_pred_lgbm = lgbm_classifier.predict(X_test_realistic_scaled)

# Evaluate the model
print("\n--- LightGBM Model Evaluation (without 'Churn Score') ---")
print("Accuracy:", accuracy_score(y_test_realistic, y_pred_lgbm))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_lgbm))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_lgbm))

### Evaluating the LightGBM Model

Now, let's critically evaluate the LightGBM model's performance and compare it against both the tuned Random Forest and the XGBoost models. The key metrics to consider are:

*   **Recall for Churn (class 1)**: How well does it identify actual churners?
*   **Precision for Churn (class 1)**: What is the proportion of correctly predicted churners among all predicted churners?
*   **F1-Score for Churn (class 1)**: How does it balance precision and recall?

This comparison will help us determine which model provides the best balance of performance for our churn prediction task, given the realistic scenario without the 'Churn Score' feature.

# 10.Exploring DL models: Simple Neural Network

Given that tree-based models (Random Forest, XGBoost, LightGBM) have shown somewhat similar performance, let's explore a different class of models: a simple Artificial Neural Network (ANN). Neural networks can learn highly complex, non-linear relationships in data that might not be easily captured by traditional models.

We will build a basic multi-layer perceptron (MLP) using `tf.keras`, which is a high-level API for TensorFlow, making it easier to construct and train neural networks. We'll also address class imbalance by using `class_weight` during training.

In [ ]:
import tensorflow as tf
from tensorflow import keras
from sklearn.utils import class_weight

print("Starting Simple Neural Network Training (without 'Churn Score')")

Starting Simple Neural Network Training (without 'Churn Score')


## 10.1 Training the neural network model

In [ ]:
# Define the neural network model
model_nn = keras.Sequential([
    keras.layers.InputLayer(input_shape=(X_train_realistic_scaled.shape[1],)), # Input layer matches number of features
    keras.layers.Dense(128, activation='relu'), # Hidden layer with 128 neurons and ReLU activation
    keras.layers.Dropout(0.3), # Dropout for regularization
    keras.layers.Dense(64, activation='relu'),  # Another hidden layer
    keras.layers.Dropout(0.3), # Dropout for regularization
    keras.layers.Dense(1, activation='sigmoid')  # Output layer with sigmoid for binary classification
])

# Compile the model
model_nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Calculate class weights for imbalance handling
class_weights = class_weight.compute_class_weight('balanced', classes=np.unique(y_train_realistic), y=y_train_realistic)
class_weights_dict = {i : class_weights[i] for i in range(len(class_weights))}

print("\nClass weights for training:", class_weights_dict)

# Train the model
history = model_nn.fit(
    X_train_realistic_scaled,
    y_train_realistic,
    epochs=50, # Number of epochs can be adjusted
    batch_size=32,
    validation_split=0.2, # Use a validation split to monitor performance during training
    class_weight=class_weights_dict,
    verbose=0 # Set verbose to 1 or 2 for more training output
)

print("\n--- Simple Neural Network Model Evaluation (without 'Churn Score') ---")

# Make predictions (probabilities)
y_pred_proba_nn = model_nn.predict(X_test_realistic_scaled)
# Convert probabilities to binary predictions
y_pred_nn = (y_pred_proba_nn > 0.5).astype(int)

# Evaluate the model
print("Accuracy:", accuracy_score(y_test_realistic, y_pred_nn))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_nn))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_nn))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/input_layer.py:27: UserWarning: Argument `input_shape` is deprecated. Use `shape` instead.
  warnings.warn(



Class weights for training: {0: np.float64(0.6775475788966514), 1: np.float64(1.9080732700135685)}

--- Simple Neural Network Model Evaluation (without 'Churn Score') ---
44/44 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
Accuracy: 0.744136460554371

Confusion Matrix:
 [[739 273]
 [ 87 308]]

Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.73      0.80      1012
           1       0.53      0.78      0.63       395

    accuracy                           0.74      1407
   macro avg       0.71      0.75      0.72      1407
weighted avg       0.79      0.74      0.76      1407



## 10.2 Optimizing the Neural Network: Hyperparameter Tuning with GridSearchCV

To further improve the performance of our simple neural network, especially its recall for the churn class, we can perform hyperparameter tuning. We'll use `GridSearchCV` from `scikit-learn` in conjunction with `KerasClassifier` (from `scikeras` which provides a scikit-learn compatible wrapper for Keras models). This allows us to systematically search for optimal network architectures and training parameters.

We will define a function to create our Keras model, allowing `GridSearchCV` to build and train models with different hyperparameter combinations.

In [ ]:
# Install a compatible version of scikeras (e.g., 0.13.0 works with scikit-learn>=1.0)
!pip install scikeras==0.13.0

import tensorflow as tf
from tensorflow import keras
from sklearn.utils import class_weight
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np

print("Starting Hyperparameter Tuning for Neural Network (without 'Churn Score')")

# Function to create the Keras model (required for KerasClassifier)
def create_nn_model(neurons_l1=128, neurons_l2=64, dropout_rate=0.3, optimizer='adam'):
    model = keras.Sequential([
        keras.layers.InputLayer(input_shape=(X_train_realistic_scaled.shape[1],)),
        keras.layers.Dense(neurons_l1, activation='relu'),
        keras.layers.Dropout(dropout_rate),
        keras.layers.Dense(64, activation='relu'),
        keras.layers.Dropout(dropout_rate),
        keras.layers.Dense(1, activation='sigmoid')
    ])
    model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])
    return model

# Wrap the Keras model in KerasClassifier
keras_clf = KerasClassifier(model=create_nn_model, verbose=0) # verbose=0 to suppress Keras training output

# Define the parameter grid to search
param_grid_nn = {
    'model__neurons_l1': [64, 128, 256],
    'model__dropout_rate': [0.2, 0.3, 0.4],
    'model__optimizer': ['adam', 'rmsprop'],
    'batch_size': [16, 32, 64],
    'epochs': [30, 50] # Reduced epochs for faster tuning, can be increased later
}

# Calculate class weights for imbalance handling
# This needs to be run before GridSearchCV.fit
class_weights = class_weight.compute_class_weight('balanced', classes=np.unique(y_train_realistic), y=y_train_realistic)
class_weights_dict = {i : class_weights[i] for i in range(len(class_weights))}


# Initialize GridSearchCV
# We'll optimize for 'recall' for the churn class (positive class, which is 1)
grid_search_nn = GridSearchCV(estimator=keras_clf, param_grid=param_grid_nn,
                              cv=3, n_jobs=-1, verbose=2, scoring='recall')

# Fit GridSearchCV
grid_search_nn.fit(X_train_realistic_scaled, y_train_realistic,
                  class_weight=class_weights_dict) # Pass class weights to fit method

print("\nBest parameters found:", grid_search_nn.best_params_)
print("Best recall score found:", grid_search_nn.best_score_)

# Get the best model
best_nn_model = grid_search_nn.best_estimator_

# Make predictions with the best model
y_pred_tuned_nn = best_nn_model.predict(X_test_realistic_scaled)

# Evaluate the tuned model
print("\n--- Tuned Neural Network Model Evaluation (without 'Churn Score') ---")
print("Accuracy: ", accuracy_score(y_test_realistic, y_pred_tuned_nn))
print("\nConfusion Matrix:\n", confusion_matrix(y_test_realistic, y_pred_tuned_nn))
print("\nClassification Report:\n", classification_report(y_test_realistic, y_pred_tuned_nn))

Found existing installation: scikeras 0.13.0
Uninstalling scikeras-0.13.0:
  Successfully uninstalled scikeras-0.13.0
  Using cached scikeras-0.13.0-py3-none-any.whl.metadata (3.1 kB)
Using cached scikeras-0.13.0-py3-none-any.whl (26 kB)


Starting Hyperparameter Tuning for Neural Network (without 'Churn Score')
Fitting 3 folds for each of 108 candidates, totalling 324 fits


KeyboardInterrupt: 

## 10.3 Export prediction csv file

In [ ]:
# Make predictions (probabilities) using the best tuned neural network model
# y_pred_proba_best_nn = best_nn_model.predict_proba(X_test_realistic_scaled)[:, 1]

# Convert probabilities to binary predictions (using 0.5 as threshold)
# y_pred_best_nn = (y_pred_proba_best_nn > 0.5).astype(int)


# Create a DataFrame to store actual and predicted values
predictions_df = pd.DataFrame({
    'Actual_Churn': y_test_realistic,
    'Predicted_Churn': y_pred_nn.flatten(), # Flatten to 1D
    'Predicted_Probability_Churn': y_pred_proba_nn.flatten() # Flatten to 1D
})

# Save the DataFrame to a CSV file
predictions_df.to_csv('nn_churn_predictions.csv', index=False)

print("Neural Network predictions saved to 'nn_churn_predictions.csv'")
display(predictions_df.head(20))

Neural Network predictions saved to 'nn_churn_predictions.csv'


,Actual_Churn,Predicted_Churn,Predicted_Probability_Churn
0,0,0,1.448053e-02
1,0,0,6.855138e-02
2,0,1,5.479431e-01
3,0,0,2.494845e-03
4,0,0,4.629608e-01
5,0,0,1.144177e-07
6,0,0,5.283220e-03
7,0,0,1.377108e-01
8,0,0,1.257171e-01
9,1,1,7.710341e-01


### Evaluating the Tuned Neural Network Model

After running the `GridSearchCV`, we will have the best set of hyperparameters for our neural network, optimized for recall. Now, let's analyze its performance:

*   **Recall for Churn (class 1)**: Has it improved significantly from the initial neural network?
*   **Precision for Churn (class 1)**: How does the precision compare, and what are the trade-offs?
*   **F1-Score for Churn (class 1)**: Does the tuned neural network offer a better balance of precision and recall?

This will help us understand if hyperparameter tuning has successfully boosted the neural network's ability to identify churners, and how it stacks up against the best performing tree-based models.

### Evaluating the Simple Neural Network Model

After training and evaluating the simple neural network, compare its performance against the tuned Random Forest, XGBoost, and LightGBM models. Pay close attention to:

*   **Recall for Churn (class 1)**: Is it better, worse, or similar to the best performing model so far?
*   **Precision for Churn (class 1)**: How does the precision compare, and what's the trade-off?
*   **F1-Score for Churn (class 1)**: Does the neural network offer a better balance of precision and recall for the churn class?

Neural networks can sometimes be more sensitive to hyperparameter choices (like the number of layers, neurons, activation functions, and learning rate) and might require more extensive tuning to reach their full potential. However, this initial run will give us an idea if this model family is a promising direction.

### Evaluating the Tuned Random Forest Model

Compare the `Classification Report` of this tuned model with the previous model (without 'Churn Score' and default hyperparameters). Pay close attention to:

*   **Recall for Churn (class 1)**: Did it improve, indicating the model is better at identifying actual churners?
*   **Precision for Churn (class 1)**: What is the trade-off with precision? Sometimes, improving recall can lead to a slight decrease in precision, and vice-versa. The `class_weight='balanced'` setting should help here.
*   **F1-Score for Churn (class 1)**: This metric balances both precision and recall. A higher F1-score indicates a better overall balance.

This iterative process of tuning and analyzing specific metrics helps in building a model that aligns with the business objectives (e.g., minimizing false negatives in churn prediction).

# 11.Conclusion

Model Performance Summary:
Our comparative analysis of classification models revealed that the Random Forest classifier achieved 93% accuracy, outperforming both Logistic Regression and SVM.

This superior performance is attributed to its ability to:
*   Capture non-linear relationships in customer behavior
*   Handle feature interactions effectively
*   Reduce overfitting through ensemble learning

Strategic Recommendations:
*   Deploy the Random Forest model for real-time churn prediction.
*   Prioritize high-risk customers (top 20% predicted churn probability) for retention efforts.


# 12.Potential Improvements

Here are some avenues to further enhance this project:

1. Advanced Hyperparameter Tuning
While you've trained the models with default or specified parameters, performing a more rigorous hyperparameter tuning (e.g., using `GridSearchCV` or `RandomizedSearchCV`) can significantly boost model performance. This is especially true for models like Random Forest, which has several parameters that can be optimized.

2. Feature Importance Analysis
For tree-based models like Random Forest, you can extract and visualize feature importances. This can provide valuable insights into which features contribute most to predicting churn, helping to understand the business drivers and potentially simplify the model by removing less important features.

3. Cross-Validation
Instead of a single train-test split, implementing k-fold cross-validation can provide a more robust evaluation of your model's performance and help in assessing its generalization capabilities more accurately.

4. Experiment with More Advanced Models
Consider exploring other powerful classification algorithms such as Gradient Boosting Machines (e.g., XGBoost, LightGBM) or even simple Neural Networks, which might capture more complex patterns in the data.

5. Error Analysis
Dive deeper into the misclassified instances. Understanding why certain customers are misclassified (e.g., false positives, false negatives) can reveal patterns or data quality issues that need to be addressed, leading to targeted feature engineering or data collection efforts.

6. Explore Different Feature Engineering Techniques
Consider creating new features from existing ones. For example:
*   **Ratio features**: Ratios of monthly charges to tenure, or other relevant combinations.
*   **Interaction terms**: Combinations of two or more features that might have a synergistic effect on churn.

7. Handling Class Imbalance
If the dataset has a significant class imbalance (more non-churners than churners), techniques like SMOTE (Synthetic Minority Over-sampling Technique) or using different class weights in models could be beneficial.

## 13. GitHub Repository Setup

To create a GitHub repository directly from this notebook, you can use the following commands. Make sure you have a GitHub account and have set up your credentials (e.g., a Personal Access Token with repo access) if you haven't already done so in your Colab environment.

First, you'll need to define your GitHub username, email, and the repository URL.

In [ ]:
# Replace with your GitHub username and email
GH_USERNAME = "your-github-username"
GH_EMAIL = "your-github-email@example.com"

# Replace with the URL of your new or existing GitHub repository
# e.g., 'https://github.com/your-github-username/your-repo-name.git'
GH_REPO_URL = "https://github.com/your-github-username/your-repo-name.git"

# Configure Git
!git config --global user.name "{GH_USERNAME}"
!git config --global user.email "{GH_EMAIL}"

# (Optional) If you're using a Personal Access Token (PAT) for authentication,
# you might need to set it up. Colab can securely store secrets.
# For example, store your PAT as a secret named 'GH_TOKEN' in Colab.
# You would then use it like this:
# !git remote set-url origin https://{GH_USERNAME}:{userdata.get('GH_TOKEN')}@github.com/{GH_USERNAME}/your-repo-name.git

print("Git configuration complete. Make sure to set up your GitHub token if required.")

Now, let's initialize the repository, add the current notebook, commit, and push. **Please ensure you have created an empty repository on GitHub first** or are pushing to an existing one.

In [ ]:
# Initialize a new Git repository (if not already done)
# This is typically done once per project folder
# !git init

# Add all files to the staging area
!git add .

# Commit the changes
!git commit -m "Initial commit: Telco Churn Prediction Notebook"

# If this is the first time you're pushing, you might need to set the upstream
# !git branch -M main
# !git remote add origin {GH_REPO_URL}

# Push to GitHub
# If you are prompted for username/password, use your GitHub username and Personal Access Token (PAT).
!git push -u origin main

print("Notebook pushed to GitHub!")

## 14. Suggestions for Improving Your Notebook for Demonstration and Sharing

Your notebook provides a good foundation for a machine learning project. To make it even more valuable, readable, and impactful for future recruiters and as a project demonstration, consider the following improvements:

### General Readability and Presentation:

1.  **Clear Executive Summary (at the top)**: Add a new Markdown cell right after the title (or at the very beginning) that provides a concise overview:
    *   **Project Goal**: What problem does this notebook solve? (e.g., predicting customer churn).
    *   **Dataset**: Briefly describe the data source and its key characteristics.
    *   **Methodology**: What techniques did you use? (e.g., EDA, feature engineering, various classification models, hyperparameter tuning).
    *   **Key Findings/Results**: What were the most important insights and the best model's performance? (e.g., "Random Forest achieved X% recall for churn, and key drivers were Y, Z...").
    *   **Business Impact/Recommendations**: What are the actionable insights or strategic recommendations based on your findings?

2.  **Consistent and Descriptive Headings**: You've done well with section headings (e.g., `## 5.1 Numerical data with churn value`). Ensure these clearly convey the content of each section.

3.  **Detailed Explanations Before Code Blocks**: Before each significant code block (especially for complex steps like preprocessing, model training, and evaluation), add a Markdown cell explaining:
    *   **What** the code is doing (e.g., "This section performs one-hot encoding on categorical features.").
    *   **Why** it's being done (e.g., "Categorical features need to be converted to numerical format for machine learning models.").
    *   **Expected Outcome** (e.g., "The resulting `X_train_ohe_df` will contain binary columns for each category.").

4.  **In-line Comments**: While explanations before blocks are good, brief, focused comments within the code itself clarify specific lines or small groups of lines.

5.  **Output Interpretation**: After running a code block that produces output (e.g., `df.info()`, `df.describe()`, classification reports, plots), add a Markdown cell that interprets the output. What insights do you gain? What decisions do these outputs lead to? (You've started this in some sections, expand it).
    *   **Example**: After `df.info()`, clearly state the data types, non-null counts, and any immediate issues observed (like 'Total Charges' being an object).

### Content and Technical Improvements:

6.  **Explicit Variable Explanation (as per your initial note)**: Your `1. Introduction` section explicitly asks for an "Explain of important variables?" but doesn't provide it. Create a new Markdown section (e.g., `1.1 Variable Description`) to explain the meaning of key columns like `Gender`, `Tenure Months`, `Monthly Charges`, `Total Charges`, `Churn Value`, `Churn Score`, `Contract`, etc. This is crucial for anyone new to the dataset.

7.  **Address the `KeyboardInterrupt`**: In your '10.2 Optimizing the Neural Network' section, the `GridSearchCV` for the Neural Network was interrupted. Either:
    *   **Re-run it successfully**: If feasible, let it complete.
    *   **Comment on it**: If it consistently runs too long, add a Markdown note explaining that it was interrupted, why, and perhaps what you would do to optimize it further (e.g., reduce parameter space, use `RandomizedSearchCV`). This shows awareness.

8.  **Comprehensive Conclusion (Section 11)**: Your current conclusion is very brief and primarily focuses on Random Forest. Expand it significantly:
    *   **Summarize ALL models**: Discuss the performance of Logistic Regression, SVM, original RF, simplified RF, XGBoost, LightGBM, and the Neural Network (both simple and tuned, if the tuning completes).
    *   **Comparative Analysis**: Directly compare the key metrics (especially recall for the churn class) across the models. Which model performed best in the *realistic scenario* (without 'Churn Score')? Why do you think so?
    *   **Trade-offs**: Discuss trade-offs (e.g., accuracy vs. interpretability, performance vs. computational cost, False Positives vs. False Negatives).
    *   **Final Recommendation**: Based on your detailed analysis, what is your ultimate recommendation for a churn prediction model, and why?
    *   **Actionable Insights**: Reiterate the most important features driving churn and what business actions could be taken.

9.  **Visualizations Enhancements**:
    *   **Meaningful Titles and Labels**: You've already done this well, but always double-check clarity.
    *   **Interpretations**: For every plot, add a markdown cell *after* the plot explaining what it shows and what conclusions can be drawn.
    *   **Interactive Plots (Optional but impactful)**: For online sharing (like GitHub), consider using libraries like Plotly or Bokeh for interactive visualizations. This is a nice-to-have.

10. **Refine 'Potential Improvements' (Section 12)**: This section is already excellent! Keep it. It shows foresight and critical thinking.

### Overall Polish:

11. **Clear Narrative Flow**: Ensure the notebook tells a cohesive story from problem definition to conclusion. Each section should logically lead to the next.

12. **Remove Redundant Code/Output**: Clean up any unnecessary print statements or repetitive code if they don't add value to the narrative.

13. **Spellcheck and Grammar**: Professional presentation matters.

By addressing these points, your notebook will not only showcase your technical skills but also your ability to communicate complex data science projects effectively, which is highly valued by recruiters.